syntax = "proto3";

package urlshort.v1;

service URLShortenerService {
  rpc CreateShortURL(CreateShortURLRequest) returns (CreateShortURLResponse);
  rpc ResolveURL(ResolveURLRequest) returns (ResolveURLResponse);
  rpc GetTopLinks(GetTopLinksRequest) returns (GetTopLinksResponse);
  rpc GetStats(GetStatsRequest) returns (GetStatsResponse);
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

message CreateShortURLRequest {
  string long_url = 1;
  optional int32 ttl_sec = 2;
  optional int32 max_clicks = 3;
  string client_ip = 4;
}

message CreateShortURLResponse {
  bool success = 1;
  string code = 2;
  string short_url = 3;
  string error = 4;
}

message ResolveURLRequest {
  string code = 1;
  bool count_click = 2;
  string client_ip = 3;
}

message ResolveURLResponse {
  int32 status = 1;
  string long_url = 2;
  string error = 3;
}

message GetTopLinksRequest {
  int32 limit = 1;
}

message GetTopLinksResponse {
  repeated LinkStats links = 1;
}

message LinkStats {
  string code = 1;
  int32 clicks = 2;
  string long_url = 3;
}

message GetStatsRequest {
  string code = 1;
}

message GetStatsResponse {
  string code = 1;
  int64 total_clicks = 2;
  int64 created_at = 3;
  bool expired = 4;
}

message HealthCheckRequest {}

message HealthCheckResponse {
  string status = 1;
  bool redis_ok = 2;
}
